src/util/futurepool.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This mod implemented a wrapped future pool that supports `on_tick()` which is driven by
+/// tasks and is invoked no less than the specific interval.
+
+use std::fmt;
+use std::cell::{Cell, RefCell, RefMut};
+use std::sync::{mpsc, Arc};
+use std::thread;
+use std::time::Duration;
+use futures::Future;
+use futures_cpupool as cpupool;
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send + 'static {
+    fn on_tick(&mut self) {}
+}
+
+#[derive(Debug)]
+/// A delegator to wrap `Context` to provide `on_tick` feature.
+struct ContextDelegator<T: Context> {
+    tick_interval: Duration,
+    /// TODO: Can be replace by `UnsafeCell` to eliminate the runtime cost of borrow checking.
+    inner: RefCell<T>,
+    last_tick: Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextDelegator<T> {
+    fn new(context: T, tick_interval: Duration) -> ContextDelegator<T> {
+        ContextDelegator {
+            tick_interval,
+            inner: RefCell::new(context),
+            last_tick: Cell::new(None),
+        }
+    }
+
+    fn get_context(&self) -> RefMut<T> {
+        self.inner.borrow_mut()
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
+        }
+        if now.duration_since(last_tick.unwrap()) < self.tick_interval {
+            return;
+        }
+        self.last_tick.set(Some(now));
+        self.get_context().on_tick();
+    }
+}
+
+/// Each `ContextDelegator` instance is invoked individually for each thread
+/// so that we mark it as Sync.
+unsafe impl<T: Context> Sync for ContextDelegator<T> {}
+
+#[derive(Debug)]
+pub struct ContextDelegators<T: Context> {
```

Comment:
I think it shouldn't be `pub`.

Reply:
Later it will be used elsewhere so that we can use the context in the future. See https://github.com/breeswish/tikv/compare/wenxuan/readpool...breeswish:wenxuan/readpool-integration#diff-a2828e4f71806f3e4623d048057803d4R589

Reply:
In my opinion, `ContextDelegators` needs to be **private** because it can only work with `FuturePool`, call any methods of it out from `FuturePool` is dangerous.

What `sync_get` really needs is `Context` but not `ContextDelegators`. I think we can adjust `ReadPool` to avoid that.